zerver/lib/streams.py

```diff
@@ -199,22 +199,33 @@ def check_for_exactly_one_stream_arg(stream_id: Optional[int], stream: Optional[
     if stream_id is not None and stream is not None:
         raise JsonableError(_("Please choose one: 'stream' or 'stream_id'."))
 
-def access_stream_for_delete_or_update(user_profile: UserProfile, stream_id: int) -> Stream:
-    # We should only ever use this for realm admins, who are allowed
-    # to delete or update all streams on their realm, even private streams
-    # to which they are not subscribed.  We do an assert here, because
-    # all callers should have the require_realm_admin decorator.
-    assert(user_profile.is_realm_admin)
+def check_stream_access_for_delete_or_update(user_profile: UserProfile, stream: Stream,
+                                             sub: Optional[Subscription]=None) -> None:
+    if stream.realm_id != user_profile.realm_id:
+        raise JsonableError(_("Invalid stream id"))
 
-    error = _("Invalid stream id")
+    if user_profile.is_realm_admin:
+        return
+
+    if sub is not None and sub.is_stream_admin:
+        return
```

Comment:
I think we need to check if the user is subscribed, and return "Invalid stream id" if it's a private stream the user isn't subscribed to; otherwise this endpoint supports checking if a private stream exists.